File: examples/speed-demo/speed_demo.cpp
// Field execution bench — wave-converted binary runs FieldX86 on the single plane.
// Convergence is Ironclad-instant at integrate; this loop measures execution only.
#include <array>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <span>
#include <string_view>
#include <vector>

#ifndef TOOLCHAIN_TAG
#define TOOLCHAIN_TAG "unknown"
#endif

namespace {

constexpr float kPhi = 0.6180339887f;
constexpr std::size_t kDieSlots = 256;
constexpr std::size_t kWaveBands = 16;
constexpr int kFramesPerEpoch = 240;
constexpr int kProgOps = 512;

enum class Op : std::uint8_t { Add, Mul, Xor, EntropyFold, WavePhase, Nop };

struct FieldInsn {
  Op op{};
  std::uint8_t dst{};
  std::uint8_t src{};
  std::uint8_t imm{};
};

[[gnu::hot]] float entropy_fold(float e, float thermo) noexcept {
  float x = e * kPhi + thermo * (1.f - kPhi);
  for (int i = 0; i < 4; ++i) {
    x = std::fma(x, 1.113f, std::sin(x * 3.14159f) * 0.01f);
  }
  return x;
}

[[gnu::hot]] float wave_phase_decouple(float phase, float speed, int band) noexcept {
  const float w = speed * static_cast<float>(band + 1) * 0.001f;
  return std::fma(std::cos(phase * w), kPhi, std::sin(phase * (1.f - kPhi)) * 0.05f);
}

[[gnu::hot]] float nexus_score(std::span<const float> weights, std::span<const float> signals) noexcept {
  float score = 0.f;
  const std::size_t n = weights.size() < signals.size() ? weights.size() : signals.size();
  for (std::size_t i = 0; i < n; ++i) {
    score = std::fma(weights[i], signals[i], score);
  }
  return score / static_cast<float>(n);
}

[[gnu::hot]] void fieldx86_run(std::span<const FieldInsn> prog, std::span<float> die) noexcept {
  for (const FieldInsn& in : prog) {
    switch (in.op) {
      case Op::Add:
        die[in.dst % kDieSlots] += die[in.src % kDieSlots];
        break;
      case Op::Mul:
        die[in.dst % kDieSlots] *= die[in.imm % kDieSlots] * 0.01f + 1.f;
        break;
      case Op::Xor:
        die[in.dst % kDieSlots] += static_cast<float>(in.imm ^ in.src) * 1e-4f;
        break;
      case Op::EntropyFold:
        die[in.dst % kDieSlots] = entropy_fold(die[in.src % kDieSlots], die[in.imm % kDieSlots]);
        break;
      case Op::WavePhase:
        die[in.dst % kDieSlots] =
            wave_phase_decouple(die[in.src % kDieSlots], die[in.imm % kDieSlots], in.imm % kWaveBands);
        break;
      case Op::Nop:
      default:
        break;
    }
  }
}

[[gnu::hot]] void run_epoch(std::span<const FieldInsn> prog, std::span<float> die,
                            std::span<float> weights, std::span<float> signals, float& sink) noexcept {
  for (int f = 0; f < kFramesPerEpoch; ++f) {
    fieldx86_run(prog, die);
    for (std::size_t b = 0; b < kWaveBands; ++b) {
      die[b] = wave_phase_decouple(die[b], die[(b + 1) % kDieSlots], static_cast<int>(b));
    }
    for (std::size_t i = 0; i < kDieSlots; ++i) {
      signals[i] = die[i] * entropy_fold(die[i], die[(i + 17) % kDieSlots]);
    }
    sink += nexus_score(weights, signals);
  }
}

int target_seconds() {
  if (const char* env = std::getenv("SPEED_DEMO_TARGET_SEC")) {
    const int v = std::atoi(env);
    if (v > 0 && v <= 600) {
      return v;
    }
  }
  return 60;
}

}  // namespace

int main() {
  const int target_sec = target_seconds();
  std::array<float, kDieSlots> die{};
  for (std::size_t i = 0; i < kDieSlots; ++i) {
    die[i] = std::sin(static_cast<float>(i) * 0.07f) * 0.5f + 0.25f;
  }

  std::vector<FieldInsn> prog;
  prog.reserve(kProgOps);
  for (int i = 0; i < kProgOps; ++i) {
    prog.push_back({static_cast<Op>(i % 6), static_cast<std::uint8_t>(i),
                    static_cast<std::uint8_t>(i * 3), static_cast<std::uint8_t>(i * 7)});
  }

  std::vector<float> weights(kDieSlots, 0.01f);
  std::vector<float> signals(kDieSlots, 0.f);
  for (std::size_t i = 0; i < kDieSlots; ++i) {
    weights[i] += static_cast<float>(i % 11) * 0.001f;
  }

  const auto t0 = std::chrono::steady_clock::now();
  const auto deadline = t0 + std::chrono::seconds(target_sec);
  auto last_tick = t0;
  int epochs = 0;
  float sink = 0.f;

  while (std::chrono::steady_clock::now() < deadline) {
    run_epoch(prog, die, weights, signals, sink);
    ++epochs;
    const auto now = std::chrono::steady_clock::now();
    const auto since_tick = std::chrono::duration<double>(now - last_tick).count();
    if (since_tick >= 1.0) {
      const double elapsed =
          std::chrono::duration<double>(now - t0).count();
      const double pct = (elapsed / static_cast<double>(target_sec)) * 100.0;
      const std::uint64_t total_ops =
          static_cast<std::uint64_t>(epochs) * static_cast<std::uint64_t>(kFramesPerEpoch) *
          static_cast<std::uint64_t>(kProgOps);
      std::cerr << "SPEED_DEMO_PROGRESS toolchain=" << TOOLCHAIN_TAG << " elapsed_sec="
                << static_cast<int>(elapsed) << " target_sec=" << target_sec
                << " pct=" << static_cast<int>(pct) << " epochs=" << epochs
                << " total_ops=" << total_ops << '\n';
      last_tick = now;
    }
  }

  const auto t1 = std::chrono::steady_clock::now();
  const double wall_ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
  const std::uint64_t total_ops = static_cast<std::uint64_t>(epochs) * kFramesPerEpoch * kProgOps;
  const double ops_per_sec = total_ops / (wall_ms / 1000.0);

  std::cout << std::fixed << std::setprecision(2)
            << "speed_demo toolchain=" << TOOLCHAIN_TAG << " target_sec=" << target_sec
            << " wall_ms=" << wall_ms << " epochs=" << epochs << " frames_per_epoch=" << kFramesPerEpoch
            << " prog_ops=" << kProgOps << " total_ops=" << total_ops
            << " ops_per_sec=" << ops_per_sec << " checksum=" << sink << '\n';
  return 0;
}

File: examples/speed-demo/speed_demo.py
#!/usr/bin/env python3
"""Field execution — same timed kernel as speed_demo.cpp (no compile in compare)."""
from __future__ import annotations

import math
import os
import sys
import time

TOOLCHAIN = os.environ.get("TOOLCHAIN_TAG", "python")
K_DIE = 256
K_WAVE = 16
K_FRAMES = 240
K_PROG = 512
K_PHI = 0.6180339887


def _fma(a: float, b: float, c: float) -> float:
    fma = getattr(math, "fma", None)
    if fma is not None:
        return fma(a, b, c)
    return a * b + c


def _finite(x: float) -> float:
    return x if math.isfinite(x) else 0.0


def _trig(fn, x: float) -> float:
    x = _finite(x)
    try:
        return fn(math.remainder(x, math.tau))
    except ValueError:
        return 0.0


def target_seconds() -> int:
    raw = os.environ.get("SPEED_DEMO_TARGET_SEC", "10")
    try:
        v = int(raw)
        if 0 < v <= 600:
            return v
    except ValueError:
        pass
    return 10


def entropy_fold(e: float, thermo: float) -> float:
    x = _finite(e) * K_PHI + _finite(thermo) * (1.0 - K_PHI)
    for _ in range(4):
        x = _finite(x)
        x = _fma(x, 1.113, _trig(math.sin, x) * 0.01)
    return _finite(x)


def wave_phase(phase: float, speed: float, band: int) -> float:
    phase = math.remainder(_finite(phase), math.tau)
    speed = _finite(speed)
    w = speed * float(band + 1) * 0.001
    return _fma(_trig(math.cos, phase * w), K_PHI, _trig(math.sin, phase * (1.0 - K_PHI)) * 0.05)


def nexus_score(weights: list[float], signals: list[float]) -> float:
    n = min(len(weights), len(signals))
    return sum(weights[i] * signals[i] for i in range(n)) / max(n, 1)


def fieldx86_run(prog: list[tuple[int, int, int, int]], die: list[float]) -> None:
    """Hot loop — local bindings + branch table (Grok16 python lane)."""
    k_die = K_DIE
    k_wave = K_WAVE
    finite = _finite
    fold = entropy_fold
    phase = wave_phase
    for op, dst, src, imm in prog:
        d = dst % k_die
        s = src % k_die
        i = imm % k_die
        if op == 0:
            die[d] = finite(die[d] + die[s])
        elif op == 1:
            die[d] = finite(die[d] * (die[i] * 0.01 + 1.0))
        elif op == 2:
            die[d] += float(imm ^ src) * 1e-4
        elif op == 3:
            die[d] = fold(die[s], die[i])
        elif op == 4:
            die[d] = phase(die[s], die[i], imm % k_wave)


def run_epoch(prog, die, weights, signals, sink: list[float]) -> None:
    for _ in range(K_FRAMES):
        fieldx86_run(prog, die)
        for b in range(K_WAVE):
            die[b] = wave_phase(die[b], die[(b + 1) % K_DIE], b)
        for i in range(K_DIE):
            signals[i] = die[i] * entropy_fold(die[i], die[(i + 17) % K_DIE])
        sink[0] += nexus_score(weights, signals)


def main() -> int:
    target = target_seconds()
    die = [math.sin(i * 0.07) * 0.5 + 0.25 for i in range(K_DIE)]
    prog = [(i % 6, i, i * 3, i * 7) for i in range(K_PROG)]
    weights = [0.01 + (i % 11) * 0.001 for i in range(K_DIE)]
    signals = [0.0] * K_DIE
    sink = [0.0]
    epochs = 0
    t0 = time.perf_counter()
    deadline = t0 + target
    last_tick = t0

    while time.perf_counter() < deadline:
        run_epoch(prog, die, weights, signals, sink)
        epochs += 1
        now = time.perf_counter()
        if now - last_tick >= 1.0:
            elapsed = now - t0
            pct = int((elapsed / target) * 100.0)
            total_ops = epochs * K_FRAMES * K_PROG
            print(
                f"SPEED_DEMO_PROGRESS toolchain={TOOLCHAIN} elapsed_sec={int(elapsed)} "
                f"target_sec={target} pct={pct} epochs={epochs} total_ops={total_ops}",
                file=sys.stderr,
            )
            last_tick = now

    wall_ms = (time.perf_counter() - t0) * 1000.0
    total_ops = epochs * K_FRAMES * K_PROG
    ops_per_sec = total_ops / (wall_ms / 1000.0) if wall_ms > 0 else 0.0
    print(
        f"speed_demo toolchain={TOOLCHAIN} target_sec={target} wall_ms={wall_ms:.2f} "
        f"epochs={epochs} frames_per_epoch={K_FRAMES} prog_ops={K_PROG} "
        f"total_ops={total_ops} ops_per_sec={ops_per_sec:.2f} checksum={sink[0]}"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())